yarnpkg/yarn
# Log network usage on installs

I took a stab at #3347, and thought it could be a useful feature for visibility in terms of what overhead does a package bring along.

**Summary**:
~~This initial implementation looks at the response header's `content-length` and aggregates them into the console reporter over the lifetime of the install.~~

This updated implementation hooks into the request manager's stream 'data' event and aggregates stream length over the lifetime of an add / install.

Looking around more closely, I found that `request-manager` is the funnel for all requests, so this is now updated to account for that.

Into the footer, it appends a new line and dynamically adjusts the size for easier human parsing :) 

<img width="571" alt="screen shot 2017-09-10 at 11 25 42 pm" src="https://user-images.githubusercontent.com/1470297/30257601-4c5b6510-9681-11e7-962e-7952ca962612.png">

**Test Plan**:

- Verify tarball size against content length (currently a discrepancy)
- Verify output against multiple package installs
- Verify output when some packages are taken from cache and others resolved

_... ^ not an exhaustive list. I'm sure there are many different combinations of install sources I'll need to consider._

I believe these tarballs should be fixtures that I'm able to compare the logged network size to, i.e. local `left-pad` fixture should be same same size as resolved `left-pad` package of the same version. 

@BYK @bestander I'd love to get some feedback on the direction ( using the console reporter to maintain state ), the emoji ⬇️ , ~~and also some direction into the other places I should look into logging network usage ( like [here](https://github.com/yarnpkg/yarn/blob/b8f8b7c3b81544e616c97c9aefc2dd909f808a65/src/util/git.js#L234) )~~

Cheers

**EDIT**: This definitely breaks all a number of snapshots. Let me know if there's a good way to verify that this maintains stability ( only showing when a flag is present, like `showPeakMemory` functions in the reporter )
